storage/tests/test_isolation.rs

```diff
@@ -122,3 +130,335 @@ fn g0_update_conflicts_fail() {
+    snapshot_2.put_val(key_1.to_owned(), ByteArray::copy(value_12.bytes()));
+    snapshot_1.put_val(key_2.to_owned(), ByteArray::copy(value_21.bytes()));
+    let read_13 = snapshot_3.get::<128>(key_1);
+    snapshot_1.commit();
+    snapshot_2.put_val(key_2.to_owned(), ByteArray::copy(value_22.bytes()));
+    let read_23 = snapshot_3.get::<128>(key_2);
+    snapshot_2.commit();
+    assert_eq!(value_10.bytes(), read_13.unwrap().bytes());
+    assert_eq!(value_20.bytes(), read_23.unwrap().bytes());
+}
+
+
+#[test]
+fn imp() {
+    init_logging();
+    let key_1: &StorageKey<'_, 48> = &StorageKey::Reference(StorageKeyReference::new(KEYSPACE_ID, ByteReference::new(&KEY_1)));
+    let key_2: &StorageKey<'_, 48> = &StorageKey::Reference(StorageKeyReference::new(KEYSPACE_ID, ByteReference::new(&KEY_2)));
+    let value_10 = ByteArray::inline([10], 1);
+    let value_20 = ByteArray::inline([20], 1);
+    let value_30 = ByteArray::inline([30], 1);
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let snapshot_setup = storage.open_snapshot_write();
+    snapshot_setup.put_val(key_1.to_owned(), ByteArray::copy(value_10.bytes()));
+    snapshot_setup.put_val(key_2.to_owned(), ByteArray::copy(value_20.bytes()));
+    let result_snapshot = snapshot_setup.commit();
+    assert!(result_snapshot.is_ok());
+
+    let snapshot_1 = storage.open_snapshot_write();
+    let snapshot_2 = storage.open_snapshot_write();
+
+    let read_11 = snapshot_1.get::<128>(key_1);
+    let to_write_11 = ByteArray::inline([read_11.unwrap().bytes()[0] + 1], 1);
+    snapshot_1.put_val(key_1.to_owned(), ByteArray::copy(to_write_11.bytes()));
+
+    let read_21 = snapshot_1.get::<128>(key_2);
+    let to_write_21 = ByteArray::inline([read_21.unwrap().bytes()[0] + 1], 1);
+    snapshot_1.put_val(key_1.to_owned(), ByteArray::copy(to_write_21.bytes()));
+
+    let read_12 = snapshot_2.get::<128>(key_1);
+    if (read_12.unwrap().bytes()[0] == 20) {
+        snapshot_2.delete(key_1.to_owned());
+    }
+    let read_22 = snapshot_2.get::<128>(key_2);
+    if (read_22.unwrap().bytes()[0] == 20) {
+        snapshot_2.delete(key_2.to_owned());
+    }
+
+    let result_1 = snapshot_1.commit();
+    let result_2 = snapshot_2.commit();
+
+    assert!(result_1.is_ok() && result_2.is_ok());
+    let reader_after = storage.open_snapshot_read();
+    let read_1_after = reader_after.get::<128>(key_1);
+    let read_2_after = reader_after.get::<128>(key_2);
+
+    let delete_went_first_1: bool = match read_1_after {
+        Some(x) => {
+            assert_eq!(value_20.bytes()[0], x.bytes()[0]);
+            true
+        }
+        None => false
+    };
+    let delete_went_first_2: bool = match read_2_after {
+        Some(x) => {
+            assert_eq!(value_30.bytes()[0], x.bytes()[0]);
+            true
+        }
+        None => true
+    };
+    assert_eq!(delete_went_first_1, delete_went_first_2);
```

Comment:
Based on the PMP from here: https://github.com/ept/hermitage/blob/master/postgres.md#predicate-many-preceders-pmp